src/pull-request/pull-request.service.ts

```diff
@@ -56,20 +56,16 @@ export class PullRequestService {
       tap(head => keyboardsHead = head),
       switchMap(() => this.gitService.readLastNote(keyboardsRepoPath)),
       map(keyboardsNoteInfo => {
-        if (
-          (new RegExp(`KDO.+${keyboardsNoteInfo.commitSha}$`).exec(singleKbNoteInfo.message) == null ||
-            new RegExp(`KDO.+${singleKbNoteInfo.commitSha}$`).exec(keyboardsNoteInfo.message) == null) &&
-          singleKbNoteInfo.message !== '' && keyboardsNoteInfo.message !== ''
-        ) {
-          throw new Error('Non-linear history. Force-push is not allowed.');
+        if (keyboardsHead !== keyboardsNoteInfo.commitSha && keyboardsNoteInfo.commitSha !== '') {
+          throw new HttpException('Keyboards repo has new changes. This is not allowed.', 409);
         }
 
-        if (keyboardsHead !== keyboardsNoteInfo.commitSha && keyboardsNoteInfo.commitSha !== '') {
-          throw new Error('Keyboards repo has new changes. This is not allowed.');
+        if (this.hasNonLinearHistory(singleKbNoteInfo, keyboardsNoteInfo)) {
+          throw new HttpException('Non-linear history in single-keyboard repo. Force-push is not allowed.', 400);
```

Comment:
Is this just for a single PR? We don't allow force push for PRs generated through KDO?

Reply:
yes, that's for a single PR and is relevant while a PR is open. 

I wrote up https://github.com/keymanapp/developer.keyman.com/issues/63 with some additional scenarios that the current implementation might not properly handle yet.